### ISSUE TITLE
Wildcard event

### ISSUE THREAD

**Issue by akashlal** [p-org/PSharp#92]
We should introduce wild card events that match against any possible type. I have two proposals. The first one addresses an immediate need while the second is more general.

First proposal is to introduce a keyword `defer_all_else`. One can use it as follows:
```
state S {
   on e1 do a1;
   on e2 do a2;
   defer_all_else;
}
```
In this case, all events except e1 and e2 (including halt) are deferred by the state.

The second proposal is to introduce a special event type called `all_else` or perhaps just `*`. One can use it in any state action (`on` action or `ignore` action or `defer`). But it can only be used once per state (no point using it more than once). For example:

```
state S {
   on e1 do a1;
   on e2 do a2;
   ignore *;
}
```
In this state, all events, not of type e1 or e2, are ignored by the state. Then `defer_all_else` is simply `defer *`.

One can argue both ways if the position of the wild card matters. That is, is `on e1 do a1; ignore *;` different from `ignore *; on e1 do a1`? I personally prefer that the wildcard match only appears as the last action on the state. 

The first proposal is relatively easy to implement. The second will take more work. What do you think?

@pdeligia @shazqadeer

**Comment by pdeligia** [comment]
Hi Akash, i think I like the 2nd proposal more than the 1st, simply because its more general and probably more flexible on the long term.

I think that `*` (i.e. `all_else`) could mean all possible events in the program, besides the ones that have already been used to declare event handlers in this particular state. In this sense, the order of `*` in `on e1 do a1; ignore *;` versus `ignore *; on e1 do a1` would not matter. I think internally (inside P#) `*` can generate a special P# event (e.g. `RemainingEvents` or something named more sensibly), which the user cannot see. When a machine is created, we already have populated the set of possible event handlers per state, so when you encounter this special event inside the Ignore/Defer/etc sets, it should simply be a matter of diffing to decide how to handle the dequeued event.

**Comment by shazqadeer** [comment]
An option that we have considered in P is to add eventsets to the language.  An eventset is a set of events.  Once an eventset has been declared (in one place), it can be used as a shorthand in many places.  This is perhaps the most general solution with the least amount of ambiguity.

**Comment by akashlal** [comment]
Even with eventsets, there is a question of having multiple actions registered on the same event. Is that an error or just that the first action wins? I like the latter.

**Comment by pdeligia** [comment]
Currently we give an error if a single event can trigger multiple actions in the same state. 

If we introduced event sets this could be a bit trickier since you could have SetA and SetB include the same event, as you said. We can easily detect this when populating the set of event handlers per state (once per machine type) and give an appropriate error.

I personally do not like much to have the first action win. This is a bit harder to implement also, since you need to extract the declaration order of the handlers (can you even do this? I am not sure if at runtime you can get the declaration order of these attributes) and store this information? This also introduces an extra thing for developers to remember. Is there any specific reason that makes you prefer this?

**Comment by akashlal** [comment]
It is easier to define an "all" event set rather than "all remaining" event set. I also find it natural to treat the list of registered actions as first-come-first-served. I think this is just a personal preference.

**Comment by akashlal** [comment]
Here is another proposal that makes it easy to support event sets:
1. We support event type inheritance. That is, if an action is registered on event type T then it will fire on all subtypes of T. 
2. Action registration is ordered: first one wins.

In this case, `defer *` is nothing but `defer Event`.

This is a deviation from the current semantics, so we have to be careful. The advantage is that we don't need a separate notion of event sets or wildcards.

**Comment by akashlal** [comment]
Another disadvantage would be that all these lookup would become O(N) instead of O(log N):

```
        private bool CanHandleEvent(Type e)
        {
            if (this.DeferredEvents.Contains(e) ||
                this.GotoTransitions.ContainsKey(e) ||
                this.PushTransitions.ContainsKey(e) ||
                this.ActionBindings.ContainsKey(e) ||
                e == typeof(GotoStateEvent))
            {
                return true;
            }

            return false;
        }
```

But considering that N is usually very small (less than 10), this perhaps does not matter. And we can amortize the cost by caching the action to be taken for each event type (which we can already do, btw).

**Comment by akashlal** [comment]
sorry, caching is difficult because there is a stack of states, not a single state.

**Comment by akashlal** [comment]
rather than changing the current semantics, a better idea is to introduce an extra attribute for event sets. For instance `OnInheritedEventDo` or `on_inherited e do foo`.

**Comment by shazqadeer** [comment]
All this is getting super-complicated.  I would like to move back to the wild-card * that represents all events and can be used only in a deferred or ignore list. Nar can achieve what he wants by deferring * and then adding a transition for the event E, F, G, etc. he really wants to handle.  

In P, we have long had a discussion about whether to enforce disjointness of event annotations locally on a state. We agreed that it should be illegal to provide two transitions on an event E; for visualization reasons we did not want to order transitions.  By extension, we also decided that it is illegal to provide two action handlers for an event E.  But is it ok to both defer E and to install an action on it or to both install an action and a transition on E?  We decided to allow it by saying that there is an order of processing.  Annotations are inherited from the state below in the stack.  On these annotations, defer annotations are applied, next action handler annotations are applied, and finally transitions are applied. The annotations that result are the final word of the behavior of the state.  If we stick with this semantics, then * can unambiguously mean "the universe of all events", yet Nar can achieve what he wants.  The design for general eventsets can be postponed.

Note that if we allow * in the ignore list, then the above outlined semantics would require reporting an error if there is an explicit action installed in the state as well.

**Comment by akashlal** [comment]
OK, I like simple. But just to understand, this is what happens in P?

```
foreach(state s in state-stack) 
{
    if(s has a transition for e) then apply the transition;
    else if(s has an action for e) then run the action;
    else if(s defers e) then defer;
    else if(s ignores e) then ignore;
}
throw UnhandledEventException();
```

**Comment by shazqadeer** [comment]
No, that is not correct, event for a machine that does not use the push transition.  Let me write down what happens for a machine that does not use the push transition.  First, ignore annotations are compiled away.  "ignore e" becomes "on e do {}".  Then the following code is executed for event handling and dequeue.
1. If an event e has been raised, goto 3.
2. Dequeue: remove the first event from the queue that is not deferred in the state; call this event e.
3. Handle: if there is a transition for e, take it else if there is an action for e, execute it, else pop the state.

An unhandled event exception is thrown whenever the state stack becomes empty.  An important fact to be noticed here is that the "defer" annotation does not affect how events are handled.  It only affects how events are dequeued.  "ignore" annotation is simply a shorthand for an action.

With push transitions, the semantics is more complicated.  The event handling code for a state depends on the states below it in the state stack.  The action handlers are inherited according to a specific inheritance logic.  These inherited handlers may be executed without popping the stack. These semantics are formally written in the P paper.  I can write them down here also but I first want to make sure that we are all on the same page with no push transitions.

**Comment by akashlal** [comment]
With this logic, then I don't understand your last comment. If one installs "defer *", then regardless of other actions installed on the state, nothing will be dequeued. How do you propose defer * be handled with the logic you have written?

**Comment by shazqadeer** [comment]
Yes, that is an issue.  I wrote my last post assuming the set of deferred events, the set of action events, and the set of transition events are disjoint.  We have to modify the story slightly if we want to allow overlap (current P does and what I wrote was a simplification of what it allows).  

We can formalize the installation details of a state as a partial map from events to either bottom, an action, or a transition.  When  a state is entered, this map is calculated as follows by applying the following steps in sequence starting from the empty map.
1. update map to put bottom against each event in a "defer" annotation.
2. update map to put installed action against each event for which one is installed.
3. update map to transition against each event on which there is a transition.

Now the sequence of steps for dequeuing and handling events is as follows:
 1. If an event e has been raised, goto 3.
 2. Dequeue: remove the first event from the queue that is not mapped to bottom in the state; call this event e.
 3. Handle: if there is an entry in the map for e, do what the entry says; otherwise pop the state.

With this map, it is now really easy to explain what happens with push transitions.  The calculated map for a state is stashed along with the state in the  state stack.  When a state is entered, to calculate its map, the starting map is the map of the state just below it on the stack. For the lowest state on the stack, the starting map is empty.

**Comment by akashlal** [comment]
Ok. In this case, is it true that if an event e is dequeued and it is not mapped in the state map (leading to a pop) then it will cause an unhandled exception because no lower state in the stack can handle it either?

**Comment by shazqadeer** [comment]
Your question made me realize that I made a mistake in my explanation.  Here is my second attempt.

We can formalize the installation details of a state as a partial map from events to either bottom or an action.  When a state is entered, this map is calculated as follows by applying the following steps in sequence starting from the empty map.
 1. update map to put bottom against each event in a "defer" annotation.
 2. update map to put installed action against each event for which one is installed.
 3. remove from map any entry for an event on which there is a transition.

Now the sequence of steps for dequeuing and handling events is as follows:
1.If an event e has been raised, goto 3.
2.Dequeue: Remove the first event from the queue that is not mapped to bottom in the state; call this event e.
3.Handle: Tf there is an outgoing transition for e, take the transition.  Otherwise, if there is an action in the map for e, perform the action.  Otherwise, pop the state.

With this map, it is now really easy to explain what happens with push transitions. The calculated map for a state is stashed along with the state in the state stack. When a state is entered, to calculate its map, the starting map is the map of the state just below it on the stack. For the lowest state on the stack, the starting map is empty.

**Comment by akashlal** [comment]
I see. Given your method, walking down the stack, one can still hit a transition, even though one cannot hit an action.
@pdeligia I don't think this is how we do things in P# right now. We should reconcile the semantics.

**Comment by akashlal** [comment]
@shazqadeer Just to confirm, lets say that a state S1 has an action for event e, S2 has a transition for e, and S3 does not have anything for e. If the state stack is [S1; S2; S3] (with S3 being the top-most) and event e arrives, then S3 will get popped and S2 will fire the transition. Correct? In other words, the action installed by S1 gets overriden by the transition in S2, thus, S3 has no action for e.

**Comment by shazqadeer** [comment]
Yes.

Sent from my Windows Phone
________________________________

**Comment by pdeligia** [comment]
Implemented in #101
